common/src/primitives/id/with_id.rs

```diff
@@ -21,7 +21,7 @@ use super::{Id, Idable};
 ///
 /// This only allows immutable access to the underlying object to prevent it from going out of sync
 /// with the ID, which is calculated for its contents. Getting an ID just returns the stored one.
-#[derive(Clone, Debug)]
+#[derive(Clone, Debug, PartialEq, Eq)]
```

Comment:
Perhaps we can manually implement PartialEq and make it only compared the id.

Reply:
This would seem to conflict with Lukas's comment above, that we want `(x == y) == (WithId(x) == WithId(y))`. unless I'm missing something...

Reply:
They should be the same, since idable by definition is implemented for things that can have a unique id. What am I missing, @iljakuklic ?

Reply:
I think Roy is referring to my earlier comment here (modifying a bit to add some context):

> Ideally, the WithId type behaves completely transparently, i.e. (WithId::new(x) == WIthId::new(y)) == (x == y) for all x and y. In other words, the == operator should behave the same on WithId as on the original value. Sadly in Transaction some fields are excluded from ID, so the property is violated [if we only compare IDs].
> 
> Conclusions:
>
> * If you want an Eq + PartialEq instance on WithId, just let it generate the auto-derived version
> * If you want to compare by IDs, use something like x.get_id() == y.get_id(), it's more explicit about what exactly is being compared.

Or did I get this wrong and `Idable` for `Transaction` does include the witness?

Reply:
Even if it doesn't include the witness, it's basically the same transaction. That's my point. The only reason we omit the witness is because it's malleable. 

This makes me think that we should treat this like we treated the BlockIndex. There's no equality there and equality should come by comparing IDs.

So now we have two options:

1. We either accept the id as a unique and identifiable thing that defines an object, and hence use id for equality by default
2. We drop the equality trait because it makes things more expensive and ambiguous, and the caller decides whether they want to compare the inner objects by getting the reference, or just use the id.

I slightly lean towards the second option because I like things being explicit. Either way, that derive has to go in my opinion. It's a bad idea.

Reply:
> So now we have two options:
> 
> 1. We either accept the id as a unique and identifiable thing that defines an object, and hence use id for equality by default
> 2. We drop the equality trait because it makes things more expensive and ambiguous, and the caller decides whether they want to compare the inner objects by getting the reference, or just use the id.
> 
> I slightly lean towards the second option because I like things being explicit. Either way, that derive has to go in my opinion. It's a bad idea.

Of these two, I'd be in favour of the second option too. We have to check to what extent is the comparison used in tests but that will be pointed out by the compiler once it's removed. The larger point is that operations for `WithId<T>` should be consistent with operations on `T` directly.